dialer.go

```diff
@@ -516,12 +552,94 @@ func (d *Dialer) Close() error {
 	close(d.closed)
 	d.lock.Lock()
 	defer d.lock.Unlock()
+	if d.domainNameTicker != nil {
+		d.domainNameTicker.Stop()
+	}
 	for _, i := range d.cache {
 		i.Close()
 	}
 	return nil
 }
 
+func (d *Dialer) pollDomainNames(ctx context.Context) {
+	type cacheEntry struct {
+		cn    instance.ConnName
+		cache monitoredCache
+	}
+
+	// List all the cache entries created with a domain name
+	d.lock.RLock()
```

Comment:
What does this lock protect? Are we sure we want to have it protect cache entries as well?

Reply:
`d.lock` (which was here before this PR) the contents of `d.cache`. Before this PR, the Dialer only added or removed single keys. Iterating through the whole cache is new behavior required by periodic DNS reconciliation.

`d.lock` only needs to protect the state of the `d.cache`. There is a separate lock on each cache entry `entry.cache.lock` which protects the internal state of each `connectionInfoCache.`